src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -10933,16 +10938,68 @@ int RGWRados::cls_bucket_list_unordered(const DoutPrefixProvider *dpp,
 
     librados::ObjectReadOperation op;
     const std::string empty_delimiter;
+
+    // Keep the previous marker to detect real progress
+    const auto prev_marker = marker;
+
     cls_rgw_bucket_list_op(op, marker, prefix, empty_delimiter,
 			   num_entries,
                            list_versions, &result);
     r = rgw_rados_operate(dpp, ioctx, oid, std::move(op), nullptr, y, 0, nullptr, &index_ver.epoch);
+    if (r == RGWBIAdvanceAndRetryError) {
+      // CLS could not return any visible entries in this round,
+      // but it advanced the marker; retry with the new marker.
+
+      // Copy marker from cls type into rgw index key (avoid type-mismatch assignment)
+      marker.name = result.marker.name;
+      marker.instance = result.marker.instance;
+
+      ldpp_dout(dpp, 10)
+        << __func__
+        << ": got RGWBIAdvanceAndRetryError"
+        << " on oid=" << oid
+        << "; prev_marker=" << prev_marker
+        << " -> marker=" << marker
+        << "; advancing marker and retrying"
+        << dendl;
+
+      // If the marker did not advance, bound the number of consecutive retries
+      if (!(prev_marker < marker)) {
+        // No forward progress; log markers, oid, and bounded retry count.
+        ldpp_dout(dpp, 10)
+          << __func__
+          << ": no marker progress"
+          << " on oid=" << oid
+          << "; prev_marker=" << prev_marker
+          << ", marker=" << marker
+          << " retries=" << (advance_only_retries + 1)
+          << "/" << MAX_ADVANCE_ONLY_RETRIES
+          << dendl;
+        if (++advance_only_retries > MAX_ADVANCE_ONLY_RETRIES) {
+          ldpp_dout(dpp, 0)
+            << "ERROR: " << __func__
+            << ": exceeded MAX_ADVANCE_ONLY_RETRIES without marker progress"
+            << " on oid=" << oid
+            << "; prev_marker=" << prev_marker
+            << ", marker=" << marker
+            << "; aborting (EIO)"
+            << dendl;
+          return -EIO;
+        }
+      } else {
```

Comment:
Help me understand what's happening between lines 10988 and 10989. We got RGWBIAdvanceAndRetryError and marker did not move beyond prev_marker, and we'll still retry. Why do we expect any advancement with the same call?

Reply:
Thank you for your feedback. I agree that there is no advance with the same marker. In this case, would it be better to terminate immediately if there is no advance with the marker, or to force an advance?

Reply:
Interesting. How would you force an advance?

Reply:
I think it should be logged with a unique string (to make searching for the condition easier in the future) and it should exit. That would at least enable us to work on the underlying issue in the future assuming there is one. I'd be curious in your testing if we ever hit that condition.

Thank you for the excellent work!